pkg/apiserver/handlers/stats/aggragator.go

```diff
@@ -0,0 +1,271 @@
+		objs = a.networkPolicyMetrics.List()
+	} else {
+		objs, _ = a.networkPolicyMetrics.ByIndex(cache.NamespaceIndex, namespace)
+	}
+
+	metrics := make([]metricsv1alpha1.NetworkPolicyMetrics, len(objs))
+	for i, obj := range objs {
+		metrics[i] = *(obj.(*metricsv1alpha1.NetworkPolicyMetrics))
+	}
+	return metrics
+}
+
+func (a *Aggregator) GetNetworkPolicyMetric(namespace, name string) (*metricsv1alpha1.NetworkPolicyMetrics, bool) {
+	obj, exists, _ := a.networkPolicyMetrics.GetByKey(namespaceKeyFunc(namespace, name))
+	if !exists {
+		return nil, false
+	}
+	return obj.(*metricsv1alpha1.NetworkPolicyMetrics), true
+}
+
+// Collect collects the node summary asynchronously to avoid the competition for the metricsLock and to save clients
+// from pending on it.
+func (a *Aggregator) Collect(nodeSummary *v1alpha1.NodeSummary) {
+	a.dataCh <- nodeSummary
+}
+
+// Run runs a loop that keeps taking metrics summary from the data channel and actually collecting them until the
+// provided stop channel is closed.
+func (a *Aggregator) Run(stopCh <-chan struct{}) {
+	klog.Info("Starting NetworkPolicy stats aggregator")
+	defer klog.Info("Shutting down NetworkPolicy stats aggregator")
+
+	klog.Info("Waiting for caches to sync for NetworkPolicy stats aggregator")
+	if !cache.WaitForCacheSync(stopCh, a.npListerSynced) {
+		klog.Error("Unable to sync caches for NetworkPolicy stats aggregator")
+		return
+	}
+	if features.DefaultFeatureGate.Enabled(features.AntreaPolicy) {
+		if !cache.WaitForCacheSync(stopCh, a.cnpListerSynced) {
+			klog.Error("Unable to sync CNP caches for NetworkPolicy stats aggregator")
+			return
+		}
+	}
+	klog.Info("Caches are synced for NetworkPolicy stats aggregator")
+
+	for {
+		select {
+		case nodeSummary := <-a.dataCh:
+			a.doCollect(nodeSummary)
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+func (a *Aggregator) doCollect(nodeSummary *v1alpha1.NodeSummary) {
+	for _, stats := range nodeSummary.NetworkPolicies {
+		switch stats.NetworkPolicy.Category {
+		case v1alpha1.NetworkPolicy:
+			obj, exists, _ := a.networkPolicyMetrics.GetByKey(namespaceKeyFunc(stats.NetworkPolicy.Namespace, stats.NetworkPolicy.Name))
+			// The policy has been removed, skip processing its metrics.
+			if !exists {
+				continue
+			}
+			curMetric := obj.(*metricsv1alpha1.NetworkPolicyMetrics)
+			// The object returned by cache is supposed to be read only, create a new object and update it.
+			newMetric := curMetric.DeepCopy()
+			addUp(&newMetric.Stats, &stats.Stats)
+			a.networkPolicyMetrics.Update(newMetric)
+		case v1alpha1.ClusterNetworkPolicy:
+			obj, exists, _ := a.clusterNetworkPolicyMetrics.GetByKey(stats.NetworkPolicy.Name)
+			if !exists {
+				continue
+			}
+			curMetric := obj.(*metricsv1alpha1.ClusterNetworkPolicyMetrics)
+			newMetric := curMetric.DeepCopy()
+			addUp(&newMetric.Stats, &stats.Stats)
+			a.clusterNetworkPolicyMetrics.Update(newMetric)
+		}
+	}
+}
+
+func namespaceKeyFunc(namespace, name string) string {
```

Comment:
i remember we also internally have another key func like this somewhere .. can we reuse it?